// ngx-ui-loader/esm2015/lib/core/ngx-ui-loader.module.js
import * as ɵngcc0 from '@angular/core';
var NgxUiLoaderModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxUiLoaderComponent } from './ngx-ui-loader.component';
import { NgxUiLoaderBlurredDirective } from './ngx-ui-loader-blurred.directive';
import { NGX_UI_LOADER_CONFIG_TOKEN } from './ngx-ui-loader-config.token';
let NgxUiLoaderModule = NgxUiLoaderModule_1 = class NgxUiLoaderModule {
    /**
     * forRoot
     * @returns A module with its provider dependencies
     */
    static forRoot(ngxUiLoaderConfig) {
        return {
            ngModule: NgxUiLoaderModule_1,
            providers: [
                {
                    provide: NGX_UI_LOADER_CONFIG_TOKEN,
                    useValue: ngxUiLoaderConfig
                }
            ]
        };
    }
};
NgxUiLoaderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxUiLoaderModule });
NgxUiLoaderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxUiLoaderModule_Factory(t) { return new (t || NgxUiLoaderModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxUiLoaderModule, { declarations: function () { return [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxUiLoaderModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective],
                exports: [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective]
            }]
    }], null, null); })();
export { NgxUiLoaderModule };
