// ngx-paypal/esm5/lib/ngx-paypal.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-paypal.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgxPaypalComponent } from './components/paypal.component';
import { ScriptService } from './services/script.service';
import { PayPalScriptService } from './services/paypal-script.service';
import * as ɵngcc0 from '@angular/core';
var NgxPayPalModule = /** @class */ (function () {
    function NgxPayPalModule() {
    }
NgxPayPalModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxPayPalModule });
NgxPayPalModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxPayPalModule_Factory(t) { return new (t || NgxPayPalModule)(); }, providers: [
        ScriptService,
        PayPalScriptService
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxPayPalModule, { declarations: function () { return [NgxPaypalComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxPaypalComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxPayPalModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    NgxPaypalComponent,
                ],
                exports: [
                    NgxPaypalComponent,
                ],
                providers: [
                    ScriptService,
                    PayPalScriptService
                ]
            }]
    }], function () { return []; }, null); })();
    return NgxPayPalModule;
}());
export { NgxPayPalModule };
