// ngx-ui-loader/esm5/lib/http/ngx-ui-loader-http.module.js
import { __decorate, __metadata, __param } from "tslib";
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { NgxUiLoaderHttpInterceptor } from './ngx-ui-loader-http.interceptor';
import { NGX_UI_LOADER_HTTP_CONFIG_TOKEN } from './ngx-ui-loader-http-config.token';
import * as ɵngcc0 from '@angular/core';
var NgxUiLoaderHttpModule = /** @class */ (function () {
    /**
     * Constructor
     */
    function NgxUiLoaderHttpModule(parentModule) {
        if (parentModule) {
            throw new Error('[ngx-ui-loader] - NgxUiLoaderHttpModule is already loaded. It should be imported in the root `AppModule` only!');
        }
    }
    NgxUiLoaderHttpModule_1 = NgxUiLoaderHttpModule;
    /**
     * forRoot
     * @returns A module with its provider dependencies
     */
    NgxUiLoaderHttpModule.forRoot = function (httpConfig) {
        return {
            ngModule: NgxUiLoaderHttpModule_1,
            providers: [
                {
                    provide: NGX_UI_LOADER_HTTP_CONFIG_TOKEN,
                    useValue: httpConfig
                }
            ]
        };
    };
    var NgxUiLoaderHttpModule_1;
    NgxUiLoaderHttpModule.ctorParameters = function () { return [
        { type: NgxUiLoaderHttpModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
    ]; };
    NgxUiLoaderHttpModule = NgxUiLoaderHttpModule_1 = __decorate([ __param(0, Optional()), __param(0, SkipSelf()),
        __metadata("design:paramtypes", [NgxUiLoaderHttpModule])
    ], NgxUiLoaderHttpModule);
NgxUiLoaderHttpModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxUiLoaderHttpModule });
NgxUiLoaderHttpModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxUiLoaderHttpModule_Factory(t) { return new (t || NgxUiLoaderHttpModule)(ɵngcc0.ɵɵinject(NgxUiLoaderHttpModule, 12)); }, providers: [
        {
            provide: HTTP_INTERCEPTORS,
            useClass: NgxUiLoaderHttpInterceptor,
            multi: true
        }
    ] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxUiLoaderHttpModule, [{
        type: NgModule,
        args: [{
                providers: [
                    {
                        provide: HTTP_INTERCEPTORS,
                        useClass: NgxUiLoaderHttpInterceptor,
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: NgxUiLoaderHttpModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();
    return NgxUiLoaderHttpModule;
}());
export { NgxUiLoaderHttpModule };
