// ngx-ui-loader/esm5/lib/core/ngx-ui-loader-blurred.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Input, OnDestroy, Renderer2, OnInit } from '@angular/core';
import { filter } from 'rxjs/operators';
import { NgxUiLoaderService } from './ngx-ui-loader.service';
import { FOREGROUND, OVERLAY_DISAPPEAR_TIME, FAST_OVERLAY_DISAPPEAR_TIME } from '../utils/constants';
import * as ɵngcc0 from '@angular/core';
var NgxUiLoaderBlurredDirective = /** @class */ (function () {
    function NgxUiLoaderBlurredDirective(elementRef, renderer, loader) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.loader = loader;
        this.blur = this.loader.getDefaultConfig().blur;
        this.loaderId = this.loader.getDefaultConfig().masterLoaderId;
        this.fastFadeOut = this.loader.getDefaultConfig().fastFadeOut;
    }
    /**
     * On Init event
     */
    NgxUiLoaderBlurredDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.showForegroundWatcher = this.loader.showForeground$
            .pipe(filter(function (showEvent) { return _this.loaderId === showEvent.loaderId; }))
            .subscribe(function (data) {
            if (data.isShow) {
                var filterValue = "blur(" + _this.blur + "px)";
                _this.renderer.setStyle(_this.elementRef.nativeElement, '-webkit-filter', filterValue);
                _this.renderer.setStyle(_this.elementRef.nativeElement, 'filter', filterValue);
            }
            else {
                setTimeout(function () {
                    if (!_this.loader.hasRunningTask(FOREGROUND, data.loaderId)) {
                        _this.renderer.setStyle(_this.elementRef.nativeElement, '-webkit-filter', 'none');
                        _this.renderer.setStyle(_this.elementRef.nativeElement, 'filter', 'none');
                    }
                }, _this.fastFadeOut ? FAST_OVERLAY_DISAPPEAR_TIME : OVERLAY_DISAPPEAR_TIME);
            }
        });
    };
    /**
     * On destroy event
     */
    NgxUiLoaderBlurredDirective.prototype.ngOnDestroy = function () {
        if (this.showForegroundWatcher) {
            this.showForegroundWatcher.unsubscribe();
        }
    };
    NgxUiLoaderBlurredDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: NgxUiLoaderService }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], NgxUiLoaderBlurredDirective.prototype, "blur", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NgxUiLoaderBlurredDirective.prototype, "loaderId", void 0);
    NgxUiLoaderBlurredDirective = __decorate([ __metadata("design:paramtypes", [ElementRef, Renderer2, NgxUiLoaderService])
    ], NgxUiLoaderBlurredDirective);
NgxUiLoaderBlurredDirective.ɵfac = function NgxUiLoaderBlurredDirective_Factory(t) { return new (t || NgxUiLoaderBlurredDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(NgxUiLoaderService)); };
NgxUiLoaderBlurredDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NgxUiLoaderBlurredDirective, selectors: [["", "ngxUiLoaderBlurred", ""]], inputs: { blur: "blur", loaderId: "loaderId" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxUiLoaderBlurredDirective, [{
        type: Directive,
        args: [{ selector: '[ngxUiLoaderBlurred]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: NgxUiLoaderService }]; }, { blur: [{
            type: Input
        }], loaderId: [{
            type: Input
        }] }); })();
    return NgxUiLoaderBlurredDirective;
}());
export { NgxUiLoaderBlurredDirective };
